### DISCUSSION TITLE
Change the list of known apps to be in CSV format

### DISCUSSION THREAD

**Discussion by lavafroth** [Universal-Debloater-Alliance/universal-android-debloater-next-generation#144] [Ideas]
The `uad_lists.json` file is currently in the JSON format wherein each entry must specify the attribute names. Since each entry has the same attributes, why not use the CSV format where the first row (header) has the column names and the following lines have the comma separated values? It should cut the file size by more than half.

**Comment by AnonymousWP** [comment]
What's the added benefit apart from reduced size?

**Reply by lavafroth** [reply]
Faster parsing. We could also probably build the objects into structs during compile time with some macros, that way the binary data will be read more efficiently.

**Reply by AnonymousWP** [reply]
Faster parsing sounds good. On the other hand: is UAD and fetching/loading the JSON-based list slow now? But as I'm not a programmer, I'm asking the opinion of our @Universal-Debloater-Alliance/developers.

**Reply by lavafroth** [reply]
The JSON fetching delay is noticeable but isn't too huge. It might increase to an undesirable level when further packages are added.

Again, this is just an idea. I'm still not sure about the best way to handle app dependency sub-lists.

**Reply by adhirajsinghchauhan** [reply]
> We could also probably build the objects into structs during compile time with some macros, that way the binary data will be read more efficiently.

UAD reads from `resources/assets/uad_lists.json` only if it can't fetch from GitHub@main. We don't want to tie the list to releases, as packages can (and should) be updated constantly, without needing someone to release a new UAD version alongside.

**Reply by adhirajsinghchauhan** [reply]
Also, while running it through some automatic JSON-to-CSV converters, I found a couple of dangling keys: `leabel`, and `labelid`.
This should've been caught at review time, whenever the offending entries were added. But it's ok for JSON, because nothing breaks and these inconsistencies don't spoil the data itself.

CSV has several downsides. One one hand, you must enforce the same order of keys->values, otherwise the entire dataset is bad. This is difficult for any human to review. Right now, in JSON, we can have them in any order and also not worry about dangling keys as above.

Another problem is if any column has a comma in it (or any separator), you have to quote it, and then also worry about which quote character you use, etc.

And with how many keys have no values at all, most rows will end up looking like this:
```csv
id,list,"multiline
description",,,,,,,,
```
Yes, so many commas at the end which looks ridiculous and makes reviewing even more difficult.

While improving perf by reducing size is good, you'd also want to make contributors' life as easy as possible, right? They shouldn't have to worry about CSV problems (same for reviewers).

**Reply by lavafroth** [reply]
Here's a middle ground:
- we create a separate branch
- for every commit, a workflow builds the json into a binary with the raw structs
- the client (GUI) fetches this blob and parses it with serde

**Reply by adhirajsinghchauhan** [reply]
Good idea, but then putting it in a branch would be a bit ugly. Internally we already debating creating a separate repo just for the list, so with this idea it makes more sense now.

**Reply by lavafroth** [reply]
A separate repo would be great. It would also allow us to properly vet app descriptions (typos, grammar, etc.) before each release.

**Reply by Frigyes06** [reply]
UAD would still refresh list at startup

**Reply by lavafroth** [reply]
Yeah, that's fine as long as it's reasonably fast.

**Comment by lavafroth** [comment]
I think we can add the aforementioned serialized blob idea to the projects section. Closing this now.

**Reply by AnonymousWP** [reply]
Go ahead when you accepted your invite to become a developer. 🙂
